File: examples/legacy/sse_client.py
#!/usr/bin/env python3
# sse_client.py
# Cliente Python para conectarse al servidor MCP Firebird usando SSE
# Para ejecutar este ejemplo: pip install requests sseclient-py

import json
import random
import string
import time
import threading
import requests
import sseclient

class McpFirebirdSseClient:
    def __init__(self, server_url='http://localhost:3003'):
        self.server_url = server_url
        self.session_id = f"python-client-{''.join(random.choices(string.ascii_lowercase + string.digits, k=10))}"
        self.request_id = 1
        self.connected = False
        self.sse_client = None
        self.event_thread = None
        self.event_handlers = {
            'message': [],
            'error': [],
            'open': []
        }
    
    def connect(self):
        try:
            print(f"Conectando a {self.server_url}...")
            
            headers = {'Accept': 'text/event-stream'}
            response = requests.get(f"{self.server_url}", stream=True, headers=headers)
            self.sse_client = sseclient.SSEClient(response)
            
            self.connected = True
            print("Conexión SSE establecida")
            
            # Notificar a los manejadores de eventos
            self._trigger_event('open')
            
            # Iniciar un hilo para procesar eventos
            self.event_thread = threading.Thread(target=self._process_events)
            self.event_thread.daemon = True
            self.event_thread.start()
            
            return True
        except Exception as e:
            print(f"Error al crear la conexión SSE: {e}")
            self._trigger_event('error', e)
            return False
    
    def _process_events(self):
        try:
            for event in self.sse_client.events():
                try:
                    data = json.loads(event.data)
                    print(f"Mensaje recibido: {data}")
                    self._trigger_event('message', data)
                except:
                    print(f"Mensaje recibido (no JSON): {event.data}")
                    self._trigger_event('message', event.data)
        except Exception as e:
            print(f"Error en el procesamiento de eventos: {e}")
            self.connected = False
            self._trigger_event('error', e)
    
    def disconnect(self):
        if self.sse_client:
            self.sse_client.close()
            self.sse_client = None
            self.connected = False
            print("Conexión SSE cerrada")
    
    def execute_method(self, method, params=None):
        if not self.connected:
            raise Exception("No hay conexión SSE activa")
        
        if params is None:
            params = {}
        
        request = {
            "jsonrpc": "2.0",
            "id": str(self.request_id),
            "method": method,
            "params": params
        }
        self.request_id += 1
        
        print(f"Enviando solicitud: {request}")
        
        response = requests.post(
            f"{self.server_url}/message?sessionId={self.session_id}",
            headers={"Content-Type": "application/json"},
            json=request
        )
        
        response_data = response.json()
        print(f"Respuesta recibida: {response_data}")
        
        return response_data
    
    def on(self, event, callback):
        if event in self.event_handlers:
            self.event_handlers[event].append(callback)
        return self
    
    def _trigger_event(self, event, data=None):
        if event in self.event_handlers:
            for callback in self.event_handlers[event]:
                callback(data)
    
    # Métodos de conveniencia para las operaciones comunes
    def list_tables(self):
        return self.execute_method("list-tables")
    
    def describe_table(self, table):
        return self.execute_method("describe-table", {"table": table})
    
    def execute_query(self, query):
        return self.execute_method("execute-query", {"query": query})
    
    def get_methods(self):
        return self.execute_method("get-methods")

# Ejemplo de uso
def main():
    client = McpFirebirdSseClient()
    
    # Registrar manejadores de eventos
    client.on('open', lambda _: print("¡Conexión abierta!"))
    client.on('message', lambda data: print(f"Nuevo mensaje: {data}"))
    client.on('error', lambda error: print(f"Error en la conexión: {error}"))
    
    try:
        if client.connect():
            # Dar tiempo para que se establezca la conexión
            time.sleep(1)
            
            # Obtener la lista de métodos disponibles
            methods = client.get_methods()
            print(f"Métodos disponibles: {methods.get('result', [])}")
            
            # Listar tablas
            tables = client.list_tables()
            print(f"Tablas disponibles: {tables.get('result', [])}")
            
            # Describir una tabla
            if tables.get('result') and len(tables['result']) > 0:
                table_info = client.describe_table(tables['result'][0])
                print(f"Información de la tabla {tables['result'][0]}: {table_info.get('result', {})}")
            
            # Ejecutar una consulta
            query_result = client.execute_query("SELECT * FROM EMPLOYEE LIMIT 5")
            print(f"Resultado de la consulta: {query_result.get('result', {})}")
            
            # Mantener la conexión abierta por un tiempo
            time.sleep(5)
    except Exception as e:
        print(f"Error: {e}")
    finally:
        client.disconnect()

if __name__ == "__main__":
    main()

File: examples/legacy/sse-client.js
// sse-client.js
// Cliente Node.js para conectarse al servidor MCP Firebird usando SSE
// Para ejecutar este ejemplo: npm install node-fetch eventsource

const fetch = require('node-fetch');
const EventSource = require('eventsource');

class McpFirebirdSseClient {
    constructor(serverUrl = 'http://localhost:3003') {
        this.serverUrl = serverUrl;
        this.eventSource = null;
        this.sessionId = `node-client-${Math.random().toString(36).substring(2, 15)}`;
        this.requestId = 1;
        this.connected = false;
        this.eventHandlers = {
            message: [],
            error: [],
            open: []
        };
    }
    
    connect() {
        return new Promise((resolve, reject) => {
            try {
                console.log(`Conectando a ${this.serverUrl}...`);
                
                this.eventSource = new EventSource(`${this.serverUrl}`);
                
                this.eventSource.onopen = () => {
                    console.log('Conexión SSE establecida');
                    this.connected = true;
                    this._triggerEvent('open');
                    resolve();
                };
                
                this.eventSource.onmessage = (event) => {
                    try {
                        const data = JSON.parse(event.data);
                        console.log('Mensaje recibido:', data);
                        this._triggerEvent('message', data);
                    } catch (error) {
                        console.log('Mensaje recibido (no JSON):', event.data);
                        this._triggerEvent('message', event.data);
                    }
                };
                
                this.eventSource.onerror = (error) => {
                    console.error('Error en la conexión SSE:', error);
                    this.connected = false;
                    this._triggerEvent('error', error);
                    reject(error);
                };
            } catch (error) {
                console.error('Error al crear la conexión SSE:', error);
                reject(error);
            }
        });
    }
    
    disconnect() {
        if (this.eventSource) {
            this.eventSource.close();
            this.eventSource = null;
            this.connected = false;
            console.log('Conexión SSE cerrada');
        }
    }
    
    async executeMethod(method, params = {}) {
        if (!this.connected) {
            throw new Error('No hay conexión SSE activa');
        }
        
        const request = {
            jsonrpc: '2.0',
            id: String(this.requestId++),
            method,
            params
        };
        
        console.log('Enviando solicitud:', request);
        
        const response = await fetch(`${this.serverUrl}/message?sessionId=${this.sessionId}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(request)
        });
        
        const responseData = await response.json();
        console.log('Respuesta recibida:', responseData);
        
        return responseData;
    }
    
    on(event, callback) {
        if (this.eventHandlers[event]) {
            this.eventHandlers[event].push(callback);
        }
        return this;
    }
    
    _triggerEvent(event, data) {
        if (this.eventHandlers[event]) {
            for (const callback of this.eventHandlers[event]) {
                callback(data);
            }
        }
    }
    
    // Métodos de conveniencia para las operaciones comunes
    async listTables() {
        return this.executeMethod('list-tables');
    }
    
    async describeTable(table) {
        return this.executeMethod('describe-table', { table });
    }
    
    async executeQuery(query) {
        return this.executeMethod('execute-query', { query });
    }
    
    async getMethods() {
        return this.executeMethod('get-methods');
    }
}

// Ejemplo de uso
async function main() {
    const client = new McpFirebirdSseClient();
    
    // Registrar manejadores de eventos
    client.on('open', () => {
        console.log('¡Conexión abierta!');
    });
    
    client.on('message', (data) => {
        console.log('Nuevo mensaje:', data);
    });
    
    client.on('error', (error) => {
        console.error('Error en la conexión:', error);
    });
    
    try {
        await client.connect();
        
        // Obtener la lista de métodos disponibles
        const methods = await client.getMethods();
        console.log('Métodos disponibles:', methods.result);
        
        // Listar tablas
        const tables = await client.listTables();
        console.log('Tablas disponibles:', tables.result);
        
        // Describir una tabla
        if (tables.result && tables.result.length > 0) {
            const tableInfo = await client.describeTable(tables.result[0]);
            console.log(`Información de la tabla ${tables.result[0]}:`, tableInfo.result);
        }
        
        // Ejecutar una consulta
        const queryResult = await client.executeQuery('SELECT * FROM EMPLOYEE LIMIT 5');
        console.log('Resultado de la consulta:', queryResult.result);
        
        // Mantener la conexión abierta por un tiempo
        await new Promise(resolve => setTimeout(resolve, 5000));
    } catch (error) {
        console.error('Error:', error);
    } finally {
        client.disconnect();
    }
}

// Ejecutar el ejemplo si este archivo se ejecuta directamente
if (require.main === module) {
    main().catch(console.error);
}

// Exportar la clase para su uso en otros archivos
module.exports = McpFirebirdSseClient;
